### PR TITLE
A new communication protocol for COMIT

### LINKED ISSUE #1827: Design a new communication protocol for COMIT
## Problem

The current communication protocol was our first attempt. Since then, we learned many things and also adopted libp2p, which made several aspects of the design redundant.

### Be "libp2p-native":

- leverage the "protocol" concept of libp2p
- follow patterns such as oneshot substreams where applicable

### Leverage the negotiation phase

We've discovered that very likely, there will always be a negotiation phase before COMIT even gets to know about the swap.
The new communication protocol should leverage this fact to simplify the communication by avoiding (re-)transmission of data the applications already know (the famous, auto-accept problem).
To solve this, the spike will need to identify, which data definitely needs to be shared among the nodes and is not in control by the applications.
We've already identified that at the very least, this includes:

- secret-hash
- transient identities (Bitcoin for now, proposal for Ethereum is available)

### Modularity & open-closed principle

At the moment, the communication protocol is a single request-response handshake.
This imposes a lot of complexity on the message format, as it has to handle all kinds of different scenarios.
Ideally, the new communication protocol has a more modular approach to allow for different messages to be sent, depending on the ledgers, assets and swap-protocols being involved.
In a way, this relates to the open-closed principle.
Extensions to the communication protocol should happen in the form of new messages instead of modifying existing ones.

### Support for more ledgers

While designing the new protocol, one should take into account the needs for support Lightning and Grin.

### PR DESCRIPTION
**Pull request by thomaseizinger**
Resolves https://github.com/comit-network/comit-rs/issues/1827.

### CODE REVIEW

File: 0023-communication-protocol.adoc
Comment (tcharding) [inline 0023-communication-protocol.adoc:82]: I don't understand the math here?  Is this the combinations of protocols that we support?  If so shouldn't it by 4, since we only support cross-chain swaps?

File: 0023-communication-protocol.adoc
Comment (tcharding) [inline 0023-communication-protocol.adoc]: May read nicer using 1-to-N instead of N-to-1 since that is the order of the English description that follows.

File: 0023-communication-protocol.adoc
Comment (tcharding) [reply 0023-communication-protocol.adoc:82]: Oh, this may be a typo.  You say 4 below.

**Review by tcharding — APPROVED** [review]
Solid work @thomaseizinger.  As per current discussion on Basecamp I treated this as a 50% review and did not touch typos/grammar or anything too trivial.  You may like to fix up the typo `s/merle/merkle/` though :)  Really thorough, was a pleasure to read.

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc]: I am not sure I understand your point here.

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc]: wrong link?

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc]: I am not sure to understand the difference with the previous section as identity are protocol-specific data.

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc:278]: Nitpick: use Javascript to allow comments.

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc]: It is a difficult question. On one hand I believe that the hash algorithm should be treated like the expiries. It can be "defaulted" and hence not present in the merkle tree or always present with the same value.

By using a multihash we are taking a characteristic of the swap and choose/specify it in a separate place than the rest of the characteristics (amounts, expiries).

Moreover, it means that we need to have added error handling logic there. I think it should all be in the Merkle tree.

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc:206]: I think this is a dangerous root as it means we cannot change the default once they are fixed (as they are not part of the merkle root, they cannot be changed).

If the default as per this communication protocol become obsolete then we would end up having to advise to always specify expiries and go against such default (as they cannot be changed).

I'd recommend to include them in the digest.

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc:373]: I am in favour of excluding identities from the digest and keep the `/comit/swap/identity` protocol defined above as unique way to communicate identities.

Either you define the `/comit/swap/identity` protocols and have identity communicated at negotiation time and set in the merkle digest, or you do not have them in the digest and exchange them via the `/comit/swap/identity` protocol.

It would unnecessary complex to have both (part of merkle root AND exchangeable via `/comit/swap/identity`.

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc]: Indeed it does, however there is no security risk to this, only downside would be 2 incompatible clients that cannot execute swaps with each others.

As we already know that it is an important piece of information then we should ensure that it is part of the RFC and kept up to date.

File: 0023-communication-protocol.adoc
Comment (D4nte) [inline 0023-communication-protocol.adoc]: One way would be to still use the merkle tree but use a generic algorithm to order the parameters that can be applied for present and future COMIT protocols.

**Review by D4nte — COMMENTED** [review]
LGTM, I gave much opinion even if it's a draft 🙃

File: 0023-communication-protocol.adoc
Comment (da-kami) [reply 0023-communication-protocol.adoc:82]: For clarification:

He means these 4:

```
HAN(Bitcoin)-HAN(Ether)
HAN(Ether)-HAN(Bitcoin)
HErc20-HAN(Bitcoin)
HAN(Bitcoin)-HErc20
```

+2 theoretical ones that do not make sense because you can just do that on Ethereum without needing atomic swaps:

```
HErc20-HAN(Ether)
HAN(Ether)-HErc20
```

File: 0023-communication-protocol.adoc
Comment (da-kami) [reply 0023-communication-protocol.adoc]: Let me try to add some information, let me know if it helps:

Cryptographic protocols may require information exchange such as e.g. the secret-hash and certain identities for HAN. But the information exchanged may differ between different cryptographic protocols. 

Thus,  the question is, how to design the **communication protocol** given different information exchange in cryptographic protocols? Thomas talks about 2 extremes, we can create the communication protocol very generic so one can pack different information into one message of one protocol, or we can split it into several protocols that have a specific concern and only one message.

File: 0023-communication-protocol.adoc
Comment (da-kami) [reply 0023-communication-protocol.adoc]: I agree this section needs some explanation why it is separate.

From a discussion I had with @thomaseizinger before he left. Yes, in the context of atomic swaps identities are protocol-specific parameters. But the identities are actually independent of the swap. They exist on the ledger and are not purely created for the swap (like e.g. the secret). Thus we could handle them separate. A protocol to exchange e.g. a Bitcoin identity could be used for all kinds of use cases, not only atomic swaps.

**Review by da-kami — COMMENTED** [review]
still in the middle of reviewing, but submitting current status.

File: 0023-communication-protocol.adoc
Comment (da-kami) [inline 0023-communication-protocol.adoc:370]: I like the proposed `announce` protocol for its simplicity.
Two comments, just for the sake of awareness: 

1. We have to very carefully select which parameters are part of the digest to avoid **collisions** (i.e. same ledgers+assets+amounts+protocol, but different identities); for a high-frequency trader it is likely to have a lot of swaps in progress. Note that this issue is a general issue that is independent of calculation of the digest, but has to be taken into account when actually designing the digest.

I assume that a (unlikely) collision (i.e. two swaps with the same digest being in the pre-announce state at the same time) is in the responsibility of the application on top of COMIT?

2. There is no way to "give feedback" to the other party any more. When sending the swap parameters we see what the other party sends and can provide feedback on what is missing/wrong. When sending a computed digest we only find a match or we fail. There is no feedback whatsoever. We can see this as advantage (more resilient to spam) and disadvantage (no feedback to a different COMIT implementation).

File: 0023-communication-protocol.adoc
Comment (da-kami) [inline 0023-communication-protocol.adoc:171]: Note: I moved my comments concerning the `announce` protocol to the Discussion Points section because they seemed to fit better there.

File: 0023-communication-protocol.adoc
Comment (da-kami) [inline 0023-communication-protocol.adoc:359]: The definition of how to calculate the digest will be per protocol-combination I suppose? I.e. one definition for HAN-HAN, one for HErc20-HAN, (...). Correct me if I got this wrong.

**Review by da-kami — APPROVED** [review]
Great work @thomaseizinger !

I added a few comments and raised questions in the "How to compute the swap digest?" section. Non-blocking.

File: 0023-communication-protocol.adoc
Comment (D4nte) [reply 0023-communication-protocol.adoc]: I think my issue is more with the wording: "Let's consider the following requirement" invite to consider an example to make the explanation clearer to the reader.

Upon reading this section, I am more confused then if the example were not here.

File: 0023-communication-protocol.adoc
Comment (D4nte) [reply 0023-communication-protocol.adoc]: Let's add more info there then please.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: What this statement aimed to make clearer is that the communication protocol is _motivated_ by us wanting to execute cryptographic protocols. Maybe this is so obvious that it doesn't need mentioning?

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:278]: I am using comments, I am using asciidoc syntax. Reader the rendered version :)

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: I am talking about the swap digest here, not the hash algorithm of the HTLC.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:206]: Did you read the whole paragraph? :sweat_smile: 

This section is meant to serve as an example of how the digest works and is not a proposal of whether the expiries should be included.
Actually, I specifically state that they should be included just 3 lines above our comment it says:

> Imagine a negotiation protocol that doesn't include expiries.

With a footnote:

> Despite the fact that they are price-relevant.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:373]: I think I agree with you on the outcome of not including identities in the digest. My reasons are different though:

If the digest where to include identities, we assume that every negotiation protocol includes them. More generally: everything that goes into the digest is assumed to be negotiated (or to have defaults in the implementation that are the same in all implementations, f.e. because they are defined in the COMIT spec)

Anyway, I think we agree that the identities should not be part of the negotiation. They conceptually belong to the execution phase.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: Solved that one by using a hash list.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: Solved that using a hash list instead.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:370]: Regarding 1)

Yes, the swap digest is not an identifier. Hence, any application using this communication protocol has to come with its own means of uniquely identifying data that flows through the system. For cnd, this likely means assigning a local ID to any request made on the REST API.

Any party receiving such an announcement message with the same digest as a previous one MUST treat them as different swaps and assign a new ID.

Regarding 2)

I would suggest to add messages for feedback later on, once we have clear needs to avoid a similar mess that we currently have with the unused decline reason.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:359]: The elements of the hash list partly depend on the protocols in use. I would suggest to define the serialization of the individual elements separately and then define per protocol, what needs to be included:

- alpha-leger
- beta-ledger
- alpha-asset
- beta-asset
- alpha-protocol
- beta-protocol
- alpha-expiry
- beta-expiry
...

Since it is a flat list, we can put whatever we want in there :)

**Comment by thomaseizinger** [comment]
@D4nte @tcharding @da-kami I've re-requested a review but also force pushed to resolve the merge conflicts. Only the last three commits are actually new - the rest is unchanged :)

File: 0023-communication-protocol.adoc
Comment (bonomat) [inline 0023-communication-protocol.adoc]: Do you mean _interactions_ ala request+response? 
Do you assume there will be an `ACK` message back or is it fire and forget (maybe libp2p offers something here already)?

File: 0023-communication-protocol.adoc
Comment (bonomat) [inline 0023-communication-protocol.adoc]: Do you mean this? 
```suggestion
In order to not fall into the same trap as with the original communication protocol, we can just opt to design libp2p protocols for things that we don't (yet) know.
```

**Review by D4nte — APPROVED** [review]

File: 0023-communication-protocol.adoc
Comment (bonomat) [inline 0023-communication-protocol.adoc:187]: Maybe unpopular opinion: why not use protobuf?

If we have _single type messages where its shape is hardcoded by the client_, what do you think of using protobuf and serialize it to binary.

File: 0023-communication-protocol.adoc
Comment (bonomat) [inline 0023-communication-protocol.adoc]: Something else comes to my mind when reading the proposed messages further down below: 

Do you foresee a `request_information` message as well? 
This could be useful in future cryptographic protocols where an actor can ask for additional information, e.g. "hey, I need a pubkey from you now to do X", "ok, now I need a second different pubkey to do Y", "I need a randomized secret hashed with sha256", ...

It's probably too early to think about this but maybe worth to keep in mind and not exclude it completely from the beginning.

File: 0023-communication-protocol.adoc
Comment (bonomat) [inline 0023-communication-protocol.adoc:221]: Let my rephrase to see if I understand the process correctly: 

* Alice adds to her cnd some swap information
* Bob does the same thing
  * both generate the `swap_digest`
* Alice sends to Bob this `swap_digest`
* If Bob knows the `swap_digest` he generates a UUIDv4 (point 3.1.2 Confirmation message) and returns it to Alice.
* ...

Isn't there a danger of a race condition where Alice sends something to Bob but Bob does not expect anything from Alice yet?

File: 0023-communication-protocol.adoc
Comment (bonomat) [inline 0023-communication-protocol.adoc:344]: I assume that finalize should involve some finale verification? 
Let's say Alice sends `finalize` to Bob, he verifies if he has all information he needs and then returns `finalize` to Alice.

File: 0023-communication-protocol.adoc
Comment (bonomat) [inline 0023-communication-protocol.adoc]: _stopped here with the review_

**Review by bonomat — COMMENTED** [review]
Good job, I like the idea of single type messages. 

@thomaseizinger : as soon as you are happy with this spike. Can you organize a meeting to quickly present the core concepts of this spike. I think everyone of the team should have a clear picture of the implications of the proposed changes before tickets end up in grooming.

File: 0023-communication-protocol.adoc
Comment (bonomat) [inline 0023-communication-protocol.adoc:356]: I understand that the swap digest is independent from the remaining messages and is used for the following things: 
* uniquely announce a swap (provided the function used to compute the digest is collision free)
* allow to verify known information against the swap digest:
  * detect if data is invalid 
  * detect if data is missing

That being said, there might be even better data structures then a multi-hash.

If the swap digest is independent of the other messages we can _play around_ with it and see what works well for us?

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: Interactions here refers to interaction with the user (like waiting for the user to send `accept` from the API).

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: Actually, I think the original is correct but this sentence uses triple negation. I will fix it, sorry for that.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:187]: I find JSON much easier to work with. If we want binary, we can use CBOR.

PS: I am not the only one disliking protobuf :)
https://reasonablypolymorphic.com/blog/protos-are-wrong/

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: Yeah I thought about this (and potentially reversing the protocols). I would follow the design of the `identity` protocol here and maybe create protocols like: `/comit/identity/bitcoin/request/1.0.0` or something like that.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:221]: Your understanding is correct.

> Isn't there a danger of a race condition where Alice sends something to Bob but Bob does not expect anything from Alice yet?

Yes that exists but I am not sure it can be solved completely.
Bob can wait for some time before rejecting the swap but it shouldn't be too long.
Generally, whatever lead the application to instruct cnd to do a swap, the same thing should happen on the other side.

File: 0023-communication-protocol.adoc
Comment (da-kami) [inline 0023-communication-protocol.adoc]:
```suggestion
Given the above protocols, it is possible to realize an implementation of COMIT that only requires a single user interaction (on both sides) before the swap starts.
```

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:344]: No, finalize doesn't involve any information.
The verification was done using the swap_digest.

Regarding the identities, I think it is reasonable to assume that they didn't mess them up.

The reasoning for the finalize message is to make sure, both parties know that there are not going to be any more messages.

**Review by da-kami — APPROVED** [review]

File: 0023-communication-protocol.adoc
Comment (bonomat) [reply 0023-communication-protocol.adoc]: Ah, misunderstood you. Thanks for the clarification.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:356]: > uniquely announce a swap (provided the function used to compute the digest is collision free)

The function is unlikely to be totally collision free because it is oriented around the "offer" that both parties gave to cnd (offer ~ swap - identities).
If we include the cnd identity of the other party within the digest, the only collisions that can happen are between those two parties. If that happens, it is unlikely to be a collision but actually an announcement of a second swap (with the same parameters).

> allow to verify known information against the swap digest:

Yes. The main purpose is to verify information without having to send a lot of data.

> That being said, there might be even better data structures then a multi-hash.

Note that the question you commented on specifically questions the use of the multihash. The alternative would be an untagged hash (where we hardcode the hash function).

> If the swap digest is independent of the other messages we can play around with it and see what works well for us?

Yes that is the idea of splitting up the messages. We should be able to change one part without affecting the others.

**Comment by thomaseizinger** [comment]
> Can you organize a meeting to quickly present the core concepts of this spike. I think everyone of the team should have a clear picture of the implications of the proposed changes before tickets end up in grooming.

@bonomat will do once it is merged!

File: 0023-communication-protocol.adoc
Comment (bonomat) [reply 0023-communication-protocol.adoc:344]: Sorry, I wasn't clear: I meant that Bob verifies that he has all information *he needs from Alice* before answering `finalize` to Alice.
 
The protocol proposed works the other way around so far: Alice tells Bob what she thinks he would need. If there is however a problem or a mismatch (an error occurred, protocol update, etc), then it might be the case that Bob would need more information from Alice. In this case, he should nor answer `finalize` to Alice.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: I've revised this section.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc]: Done!

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:370]: I added a section about collision resistance.

File: 0023-communication-protocol.adoc
Comment (bonomat) [reply 0023-communication-protocol.adoc:221]: Should Bob in this case indicate that he doesn't know what Alice is talking about? Meaning, return something?

**Review by tcharding — APPROVED** [review]

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:344]: > Alice tells Bob what she thinks he would need.

Not quite: Both parties tell the other parties what they think they will need.

> Sorry, I wasn't clear: I meant that Bob verifies that he has all information he needs from Alice before answering finalize to Alice.

This is what should happen yes.
The idea is that each party doesn't send `finalize` until:

- they sent all the data they intend to send
- they received all the data they need

If the finalize handshake doesn't happen within a certain time period, we should timeout and abort the swap (there seems to be a bug in the implementations then).

Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:344]: Note that the new protocol doesn't have request-reply but just one-way messages. They also all happen on different substreams, meaning it all happens concurrently.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:221]: For example?

A string that Alice can print to the logs?
Should she be able to automatically recover from that?

That topic is a whole new can of worms.
My personal preference would be to evaluate that the need for this as we go and use it ourselves and have other people use it.
That should tell us about common error scenarios and we can think about how to attack them.

I see the general problem, but I don't have a feasible solution at this stage.

**Comment by thomaseizinger** [comment]
I think all of the points are resolved, I will merge this and organize a presentation meeting for next week.

Happy to answer more questions there.

File: 0023-communication-protocol.adoc
Comment (thomaseizinger) [reply 0023-communication-protocol.adoc:187]: I just realized that you commented on the digest computation here.

I wouldn't use JSON for that but neither protobuf. The serialization is going to be very simple, like a single key-value which is then hashed.

An example representation could be:

```rust
format!("alpha_ledger={}", alpha_ledger)
```

and then hashing it.

I don't see a point in using a technology like protobuf here, we don't have any structs or anything which kind of defeats the point of an IDL.

Note: Very likely, we don't want to go with the above example because it is string-encoding dependent but rather define our own binary encoding.

**Pull request merged**
